consensus/core/src/leader_timeout.rs

```diff
@@ -0,0 +1,208 @@
+use std::sync::Arc;
+use std::time::Duration;
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+use crate::block::Round;
+use crate::context::Context;
+use crate::core::CoreSignalsReceivers;
+use crate::core_thread::CoreThreadDispatcherInterface;
+use tokio::sync::oneshot::{Receiver, Sender};
+use tokio::task::JoinHandle;
+use tokio::time::{sleep_until, Instant};
+use tracing::{debug, warn};
+
+pub(crate) struct LeaderTimeoutTaskHandle {
+    handle: JoinHandle<()>,
+    stop: Sender<()>,
+}
+
+impl LeaderTimeoutTaskHandle {
+    pub async fn stop(self) {
+        self.stop.send(()).ok();
+        self.handle.await.ok();
+    }
+}
+
+pub(crate) struct LeaderTimeoutTask<D> {
```

Comment:
We _might_ need to add more logic, but primarily it's because there wasn't too many options anyways. Ideally we would like to build this in the `Core` component, but `Core` is accessed via the `CoreThreadDispatcher`, so things would start becoming a bit complex.